plugins/test-runners/pytest/cosmic_ray_pytest_runner/runner.py

```diff
@@ -29,15 +27,31 @@ class PytestRunner(TestRunner):
     """
 
     def _run(self):
+        try:
+            from StringIO import StringIO
+        except ImportError:
+            from io import StringIO
         collector = ResultCollector()
 
+        class TestRunnerFailure(Exception):
+            def __init__(self, msg, exit_code, output):
+                self.msg = msg
+                self.exit_code = exit_code
+                self.output = output
+
         args = self.test_args
         if args:
             args = args.split()
         else:
             args = []
-        with open(os.devnull, 'w') as devnull, redirect_stdout(devnull):
-            pytest.main(args, plugins=[collector])
+        stdout = StringIO()
```

Comment:
The `StringIO` object is never `close()`d – this isn't super important, but it's good practice rather than relying on the garbage collect to dispose of the buffer. We should probably manage so `stdout` object using its context-manager interface, so:

`with StringIO as stdout:`